### PR TITLE
Adds getCircleBitmap method in GCMIntentService.java to be used in setNotificationLargeIcon

### PR DESCRIPTION
**Pull request by bmwertman**
## Description
Adds method `getCircleBitmap` and calls it within `setNotificationLargeIcon` passing in either the bitmap created from image URL or the image retrieved from local assets. Returns a round bitmap.

## Related Issue
#1464 

## Motivation and Context
Large icons in Android push notifications are by default round. The plugin prior to introducing the `getCircleBitmap` method created a square icon. `getCircleBitmap` converts all large icon images to a round bitmap prior to inserting them keeping with the default large icon style. 

## How Has This Been Tested?
Checked to ensure app handled notifications from closed, background and foreground as expected.
Tested on the following device & environment;
#### Platform and Version
     Android 5.0.2

#### What device vendor 
    Oneplus One

#### Cordova CLI version and cordova platform version
    cordova --version                          6.4
    cordova platform version android           6.1
    cordova platform version ios           4.3

#### Plugin version
    cordova plugin version | grep phonegap-plugin-push   1.9.1 with pulled changes from 1.9.2

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

**Comment by macdonst** [comment]
@bmwertman no, circular icons are not the default. By convention if you are showing a picture of a person google tends to use a circle but for other images it recommends rectangles. You can see that in their material design guidelines:

https://material.io/guidelines/patterns/notifications.html#

Not to say this isn't a good bonus feature for the plugin.

However, you mention you copied the code from a blog post. Did you get permission from the blog post owner to use that code? If the code does not have a license associated with it then it is assumed to be copywritten by the author of the post. I don't want to run into any legal issues.

Finally, I don't like the fact that it is always a circular icon. That should be an option.

File: src/android/com/adobe/phonegap/push/GCMIntentService.java [hunk 1]
```diff
@@ -690,19 +696,43 @@ private void setNotificationPriority(Bundle extras, NotificationCompat.Builder m
...
+
     private void setNotificationLargeIcon(Bundle extras, String packageName, Resources resources, NotificationCompat.Builder mBuilder) {
         String gcmLargeIcon = extras.getString(IMAGE); // from gcm
         if (gcmLargeIcon != null && !"".equals(gcmLargeIcon)) {
             if (gcmLargeIcon.startsWith("http://") || gcmLargeIcon.startsWith("https://")) {
-                mBuilder.setLargeIcon(getBitmapFromURL(gcmLargeIcon));
+                Bitmap bitmap = getBitmapFromURL(gcmLargeIcon);
+                Bitmap bm = getCircleBitmap(bitmap);
+                mBuilder.setLargeIcon(bm);
                 Log.d(LOG_TAG, "using remote large-icon from gcm");
```
Comment (macdonst) [inline src/android/com/adobe/phonegap/push/GCMIntentService.java]: I'd like to see the above code be something like this, right above the if statement add:

```
boolean roundIcon = extras.getBoolean(ROUND_ICON):
```

of course `ROUND_ICON` will need to be added to the constants file.

then inside the if/else you could do:

```
if (roundIcon) {
  mBuilder.setLargeIcon(getCircleBitmap(getBitmapFromURL(gcmLargeIcon)));
} else {
  mBuilder.setLargeIcon(getBitmapFromURL(gcmLargeIcon)); 
}
```

File: src/android/com/adobe/phonegap/push/GCMIntentService.java [hunk 2]
```diff
@@ -690,19 +696,43 @@ private void setNotificationPriority(Bundle extras, NotificationCompat.Builder m
...
             } else {
                 AssetManager assetManager = getAssets();
                 InputStream istr;
                 try {
                     istr = assetManager.open(gcmLargeIcon);
                     Bitmap bitmap = BitmapFactory.decodeStream(istr);
-                    mBuilder.setLargeIcon(bitmap);
+                    Bitmap bm = getCircleBitmap(bitmap);
+                    mBuilder.setLargeIcon(bm);
                     Log.d(LOG_TAG, "using assets large-icon from gcm");
```
Comment (macdonst) [inline src/android/com/adobe/phonegap/push/GCMIntentService.java]:
```
if (roundIcon) {
  mBuilder.setLargeIcon(getCircleBitmap(bitmap);
} else {
  mBuilder.setLargeIcon(bitmap); 
}
```

Comment (bmwertman) [inline src/android/com/adobe/phonegap/push/GCMIntentService.java]: Cool. I'm fine with making it an option. 

As far back as I've had Android devices and my memory serves, not sure which is longer, all large notifications have been round. The revision you have above is very similar to what I had at commit `9c351c2`. I was just having trouble getting the condition to catch that I had passed `roundIcon: true` to it. Would you want to see the condition set in `init` as @fr` suggested? Or do you like it as part of the ` data` payload?

Regarding copyright, I hadn't asked permission. Some experience with a former cofounder of mine led me into a discussion with a patent attorney. Basically only creative works are defensible as copyright. I wouldn't say something as standard as creating a circular image from a square one with an API falls within that realm. But I'm happy to post a question on the blog post and put a screenshot of the author's reply here if you like.

**Comment by fredgalvao** [comment]
```
Large icons should be circular when showing a person, but square in all other cases.
```

https://material.io/guidelines/patterns/notifications.html#notifications-style

**Comment by macdonst** [comment]
@fredgalvao yeah, I mentioned that in my comments as well.

**Comment by macdonst** [comment]
@bmwertman you still working on this PR? I think it is a cool new feature.

**Comment by bmwertman** [comment]
@macdonst I got a little side tracked in the holidays but I agree. I've got some things to get caught up on this week but I think I can get to it next week.

**Comment by macdonst** [comment]
@bmwertman that's cool. I actually took time off during the holidays too. I was actually too hard to get to my laptop with all the family around. I'm just going over all the activity I missed and wanted to ping you to see if there is still interest. Since there is I'll leave it to you and merge when you are ready.

**Comment by macdonst** [comment]
@bmwertman I cleaned up your code a bit and committed it as: https://github.com/phonegap/phonegap-plugin-push/commit/0dd0d468868f0b13c8d840c78dd89fca5920cd32

Basically I find the center of the image and draw a circle instead of an oval. It gets around the issue if the width and height are not equal. Otherwise you can get a shape that looks a lot like an egg. I also added a new parameter so you can tell the large icon to be displayed as a square (default) or circle.

**Pull request closed without merging**